ELClient/ELClientWebServer.cpp

```diff
@@ -0,0 +1,407 @@
+void ELClientWebServer::setup()
+{
+  registerCallback();
+}
+
+void ELClientWebServer::registerHandler(const char * URL, WebServerCallback callback)
+{
+  String s = URL;
+  registerHandler(s, callback);
+}
+
+void ELClientWebServer::registerHandler(const __FlashStringHelper * URL, WebServerCallback callback)
+{
+  String s = URL;
+  registerHandler(s, callback);
+}
+
+// registers a new handler
+void ELClientWebServer::registerHandler(const String &URL, WebServerCallback callback)
+{
+  unregisterHandler(URL);
+  
+  struct _Handler * hnd = new struct _Handler(); // "new" is used here instead of malloc to call String destructor at freeing. DOn't use malloc/free.
+  hnd->URL = URL;             // handler URL
+  hnd->callback = callback;   // callback
+  hnd->next = handlers;       // next handler
+  handlers = hnd;             // change the first handler
+}
+
+void ELClientWebServer::unregisterHandler(const char * URL)
+{
+  String s = URL;
+  unregisterHandler(s);
+}
+
+void ELClientWebServer::unregisterHandler(const __FlashStringHelper * URL)
+{
+  String s = URL;
+  unregisterHandler(s);
+}
+
+// unregisters a previously registered handler
+void ELClientWebServer::unregisterHandler(const String &URL)
+{
+  struct _Handler *prev = 0;
+  struct _Handler *hnd = handlers;
+  while( hnd != 0 )
+  {
+    if( hnd->URL == URL )
+    {
+      if( prev == 0 )
+        handlers = hnd->next;
+      else
+        prev->next = hnd->next;
+      
+      delete hnd;
+    }
+    prev = hnd;
+    hnd = hnd->next;
+  }
+}
+
+void ELClientWebServer::registerCallback()
+{
+  // WebServer doesn't send messages to MCU only if asked
+  // register here to the web callback
+  // periodic reregistration is required in case of ESP8266 reset
+  _elc->Request(CMD_CB_ADD, 100, 1);
+  _elc->Request(F("webCb"), 5);
+  _elc->Request();
+}
+
+void ELClientWebServer::processPacket(ELClientPacket *packet)
+{
+  ELClientResponse response(packet);
+
+  uint16_t shrt;
+  response.popArg(&shrt, 2);
+  RequestReason reason = (RequestReason)shrt; // request reason
+
+  response.popArg(remote_ip, 4);    // remote IP address
+  response.popArg(&remote_port, 2); // remote port
+  
+  char * url;
+  int urlLen = response.popArgPtr(&url);
+  
+  WebServerCallback callback = 0;
+  
+  struct _Handler *hnd = handlers;
+  while( hnd != 0 )
+  {
+    if( hnd->URL.length() == urlLen && memcmp( url, hnd->URL.begin(), urlLen ) == 0 )
+    {
+      callback = hnd->callback;
+      break;
+    }
+    hnd = hnd->next;
+  }
+
+  if( hnd == 0 ) // no handler found for the URL
+  {
+    _elc->_debug->print(F("Handler not found for URL:"));
+    
+    for(int i=0; i < urlLen; i++)
+      _elc->_debug->print( url[i] );
+    _elc->_debug->println();
+    return;
+  }
+  
+  switch(reason)
+  {
+    case WS_BUTTON: // invoked when a button pressed
+      {
+        char * idPtr;
+        int idLen = response.popArgPtr(&idPtr);
+  
+        char id[idLen+1];
+        memcpy(id, idPtr, idLen);
```

Comment:
The terminating 0 is missing. The final solution would be very-very complex. Do this here, add the terminating 0, call the callback.